FWCore/Integration/test/ConcurrentIOVESSource.cc

```diff
@@ -0,0 +1,193 @@
+// -*- C++ -*-
+//
+// Package:     FWCore/Integration
+// Class  :     ConcurrentIOVESSource
+//
+// Implementation:
+//     ESSource used for tests of Framework support for
+//     concurrent IOVs in the EventSetup system
+//
+// Original Author:  W. David Dagenhart
+//         Created:  21 March 2019
+
+#include "DataFormats/Provenance/interface/EventID.h"
+#include "DataFormats/Provenance/interface/Timestamp.h"
+#include "FWCore/Framework/interface/EventSetupRecordIntervalFinder.h"
+#include "FWCore/Framework/interface/ESProducer.h"
+#include "FWCore/Framework/interface/IOVSyncValue.h"
+#include "FWCore/Framework/interface/SourceFactory.h"
+#include "FWCore/Framework/interface/ValidityInterval.h"
+#include "FWCore/Integration/interface/ESTestData.h"
+#include "FWCore/Integration/interface/ESTestRecords.h"
+#include "FWCore/Integration/test/IOVTestInfo.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Utilities/interface/WallclockTimer.h"
+
+#include <memory>
+#include <vector>
+
+namespace edmtest {
+
+  class ConcurrentIOVESSource : public edm::EventSetupRecordIntervalFinder, public edm::ESProducer {
+  public:
+    ConcurrentIOVESSource(edm::ParameterSet const&);
+
+    std::unique_ptr<IOVTestInfo> produce(ESTestRecordI const&);
+    std::unique_ptr<ESTestDataA> produceA(ESTestRecordA const&);
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void setIntervalFor(edm::eventsetup::EventSetupRecordKey const&,
+                        edm::IOVSyncValue const&,
+                        edm::ValidityInterval&) override;
+
+    bool isConcurrentFinder() const override { return concurrentFinder_; }
+
+    // These are thread safe because after the constructor we do not
+    // modify their state.
+    bool iovIsTime_;
```

Comment:
You can make these `const` since they are set in the initialization list.

Reply:
Done.